src/query_plan/operation.rs

```diff
@@ -0,0 +1,435 @@
+};
+use apollo_compiler::{Node, Schema};
+use indexmap::IndexMap;
+
+// copy of apollo compiler types that store selections in a map so we can normalize it efficiently
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NormalizedSelectionSet {
+    pub ty: NamedType,
+    pub selections: NormalizedSelectionMap,
+}
+
+pub type NormalizedSelectionMap = IndexMap<NormalizedSelectionKey, NormalizedSelection>;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum NormalizedSelectionKey {
+    Field {
+        name: Name,
+        directives: DirectiveList,
+    },
+    InlineFragment {
+        type_condition: Option<Name>,
+        directives: DirectiveList,
+    },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum NormalizedSelection {
+    NormalizedField(Node<NormalizedField>),
+    NormalizedInlineFragment(Node<NormalizedInlineFragment>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NormalizedField {
+    pub definition: Node<FieldDefinition>,
+    pub alias: Option<Name>,
+    pub name: Name,
+    pub arguments: Vec<Node<Argument>>,
+    pub directives: DirectiveList,
+    pub selection_set: NormalizedSelectionSet,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NormalizedInlineFragment {
+    pub type_condition: Option<NamedType>,
+    pub directives: DirectiveList,
+    pub selection_set: NormalizedSelectionSet,
+}
+
+impl NormalizedSelectionSet {
+    // cannot use From trait as we need to pass the fragments
+    // otherwise we need two passes - one to change to map and another one to expand fragments
+    fn from_selection_set(
+        selection_set: &SelectionSet,
+        fragments: &IndexMap<Name, Node<Fragment>>,
+    ) -> Self {
+        let normalized_selections = normalize_selections(&selection_set.selections, fragments);
+        NormalizedSelectionSet {
+            ty: selection_set.ty.clone(),
+            selections: normalized_selections,
+        }
+    }
+}
+
+impl From<NormalizedSelectionSet> for SelectionSet {
+    fn from(val: NormalizedSelectionSet) -> Self {
+        SelectionSet {
+            ty: val.ty.clone(),
+            selections: flatten_selections(&val.selections),
+        }
+    }
+}
+
+/// Converts vec of Selections to a map of NormalizedSelections
+///
+/// Expands all named fragments and merge duplicate selections.
+fn normalize_selections(
+    selections: &Vec<Selection>,
+    fragments: &IndexMap<Name, Node<Fragment>>,
+) -> NormalizedSelectionMap {
+    let mut normalized = NormalizedSelectionMap::new();
+    for selection in selections {
+        match selection {
+            Selection::Field(field) => {
+                let expanded_selection_set =
+                    normalize_selections(&field.selection_set.selections.to_owned(), fragments);
+
+                if let NormalizedSelection::NormalizedField(field_entry) = normalized
+                    .entry(field.into())
+                    .or_insert(NormalizedSelection::NormalizedField(Node::new(
```

Comment:
A minor nit, but if you use `.or_insert_with(|| { ... })` here instead of `.or_insert( ... )`, you can avoid the cost of creating the `NormalizedSelection::NormalizedField` when the entry is already present.